### DISCUSSION TITLE
ConfigValidationError sklearn-cat

### DISCUSSION THREAD

**Discussion by yw2903** [explosion/spaCy#11622] [Help: Coding & Implementations] [answered]
I got the following error when loading a nlp that has been trained and exported out to disk.
```
ConfigValidationError: 

Config validation error

sklearn-cat -> sklearn_model   field required
sklearn-cat -> label           field required
sklearn-cat -> classes         field required

{'nlp': <spacy.lang.en.English object at 0x7f07d879e6a0>, 'name': 'sklearn-cat', '@factories': 'sklearn-cat'}
```
And here is how I train the model (it works fine right after I train it, the issue is reloading the nlp model after exporting it)
```
X = [
    'i really like this post',
    'thanks for that comment',
    'i enjoy this friendly forum',
    'this is a bad post',
    'i dislike this article',
    'this is not well written'
]

y = ['pos', 'pos', 'pos', 'neg', 'neg', 'neg']
pipe = make_pipeline(CountVectorizer(), LogisticRegression())

pipe.fit(X,y)

nlp = spacy.load("en_core_web_sm")

@Language.component("sklearn-cat")
def my_component(doc):
    pred = pipe.predict([doc.text])[0]
    proba = pipe.predict_proba([doc.text]).max()
    doc.cats[pred] = proba
    return doc
#conf = {"sklearn_model": "@sklearn_model_basic.v1", "label": "pos", "classes": ["pos", "neg"]}
nlp.add_pipe("sklearn-cat")
nlp("can you flip a coin?").cats
```
Suppose I add that config line, I will experience the following error:
```
Config validation error

sklearn-cat -> classes         extra fields not permitted
sklearn-cat -> label           extra fields not permitted
sklearn-cat -> sklearn_model   extra fields not permitted

{'nlp': <spacy.lang.en.English object at 0x7f9d959f08e0>, 'name': 'sklearn-cat', 'sklearn_model': '@sklearn_model_basic.v1', 'label': 'pos', 'classes': ['pos', 'neg'], '@factories': 'sklearn-cat'}
```

**Comment by ljvmiranda921** [comment, marked as answer]
Hi @yw2903 ,

There are two ways you can solve this:

1. A minimal correction might be to use [`Language.factory`](https://spacy.io/api/language#factory) instead of [`Language.component`](https://spacy.io/api/language#component), then include the extra settings in the `my_component` function (c.f. https://spacy.io/usage/processing-pipelines#custom-components-factories). Something like this:

```python
@Language.factory("sklearn-cat")
def my_component(doc, sklearn_model: str, label: str, classes: List[str])):
    # do something
    return doc
```

This might work as a toy example but it's not the best practice (harder to extend, you have something happening in the global namespace, etc.). Ideally you want everything to happen inside a single class, it's more extensible that way. And so, another approach would be to...

2. Create a factory with a full class (you can check [this example](https://spacy.io/usage/processing-pipelines#example-stateful-components)). In this approach, it's a good idea to (1) fit the sklearn model in `__init__` and then (2) run the model in `__call__`. Again, you can check the linked example as a starting point.

**Reply by yw2903** [reply]
Thanks for the response.

I got the following error when switching to **Language.factory**

```
ValueError: [E004] Can't set up pipeline component: a factory for 'sklearn-cat' already exists. Existing factory: <function my_component at 0x7f9859d94f70>. New factory: <function my_component at 0x7f9859bdcca0>
```

**Reply by yw2903** [reply]
@ljvmiranda921 
I still trouble transferring the model using the Example you pointed to. I run the example and exported using nlp.to_disk, then I got the following error loading it back in:

```
ValueError: [E002] Can't find factory for 'acronyms' for language English (en). This usually happens when spaCy calls `nlp.create_pipe` with a custom component name that's not registered on the current language class. If you're using a Transformer, make sure to install 'spacy-transformers'. If you're using a custom component, make sure you've added the decorator `@Language.component` (for function components) or `@Language.factory` (for class components).

Available factories: attribute_ruler, tok2vec, merge_noun_chunks, merge_entities, merge_subtokens, token_splitter, doc_cleaner, parser, beam_parser, entity_linker, ner, beam_ner, entity_ruler, lemmatizer, tagger, morphologizer, senter, sentencizer, textcat, spancat, textcat_multilabel, en.lemmatizer
```

**Reply by ljvmiranda921** [reply]
Hi @yw2903  , just to sanity-check, are you [passing the custom code](https://spacy.io/api/cli#train) when you run `spacy train` or `evaluate`? You can pass the path to the Python file that contains this custom component using the `--code` (or `-c`) parameter.

**Reply by yw2903** [reply]
I'm running the code in Jupyter Notebook

**Reply by ljvmiranda921** [reply]
Hi! It might be helpful (and more extensible) to write the custom component in a separate Python file, then call the `train` command with the `--code` parameter.
